mmdet/evaluation/metrics/coco_panoptic_metric.py

```diff
@@ -226,142 +236,282 @@ def result2json(self, results: Sequence[dict],
+                    mask = pan_png == id
+                    isthing = categories[label]['isthing']
+                    if isthing:
+                        iscrowd = 1 if not segment_info['is_thing'] else 0
+                    else:
+                        iscrowd = 0
+
+                    new_segment_info = {
+                        'id': id,
+                        'category_id': label,
+                        'isthing': isthing,
+                        'iscrowd': iscrowd,
+                        'area': mask.sum()
+                    }
+                    segments_info.append(new_segment_info)
+            else:
+                # get segments_info from annotation file
+                segments_info = self._coco_api.imgToAnns[img_id]
+
+            gt['segments_info'] = segments_info
+
+            pq_stats = pq_compute_single_core(
+                proc_id=0,
+                annotation_set=[(gt, result)],
+                gt_folder=self.seg_prefix,
+                pred_folder=self.seg_out_dir,
+                categories=categories,
+                file_client=self.file_client)
+
+            self.results.append(pq_stats)
+
+    def _process_without_tmpfile(self, data_batch: Sequence[dict],
+                                 predictions: Sequence[dict]):
+        """Process gts and predictions when ``outfile_prefix`` is set.
+
+        The predictions will be saved to directory specified by
+        ``outfile_predfix``. The matched pair (gt, result) will be put into
+        ``self.results``.
+        """
+        for data, pred in zip(data_batch, predictions):
+            # parse pred
+            img_id = data['data_sample']['img_id']
+            segm_file = osp.basename(data['data_sample']['img_path']).replace(
+                'jpg', 'png')
+            result = self._parse_predictions(
+                pred=pred, img_id=img_id, segm_file=segm_file)
+
+            # parse gt
+            gt = dict()
+            gt['image_id'] = img_id
+            gt['width'] = data['data_sample']['ori_shape'][1]
+            gt['height'] = data['data_sample']['ori_shape'][0]
+
+            if self._coco_api is None:
+                # get segments_info from dataset
                 gt['segments_info'] = data['data_sample']['segments_info']
                 gt['seg_map_path'] = data['data_sample']['seg_map_path']
 
             self.results.append((gt, result))
 
+    def process(self, data_batch: Sequence[dict],
+                predictions: Sequence[dict]) -> None:
+        """Process one batch of data samples and predictions. The processed
+        results should be stored in ``self.results``, which will be used to
+        compute the metrics when all batches have been processed.
+
+        Args:
+            data_batch (Sequence[dict]): A batch of data
+                from the dataloader.
+            predictions (Sequence[dict]): A batch of outputs from
+                the model.
+        """
+        if self.tmp_dir is None:
```

Comment:
add comments here that when `tmp_dir is not None`, it will compute pq_stats here, otherwise, it will save gt and predictions to self.results.